RFC: Remove boost usage from kernel api / headers

sedited opened the pull request:
Similarly to #28327 I wanted to open this PR to receive some opinions and better approaches.

The kernel library is currently at the stage where unwanted headers are removed from its set of headers. In practice, this means we are reducing the number of includes that are required for compiling the experimental `bitcoin-chainstate` binary. This is described in stage 1 step 3 of the [project tracking issue](https://github.com/bitcoin/bitcoin/issues/27587).

Currently the mempool is part of the kernel library. The mempool headers include the boost multi index headers. Thus any application wanting to use the kernel library and its headers will have to include the boost headers too. This is not only undesirable because of the sheer size of these headers, but also might lead to conflicts if the including application uses a different boost version.

In the approach laid out by this PR, mempool member variables and methods are declared in the header without having to include boost by either wrapping them in a struct and [pimpling](https://en.cppreference.com/w/cpp/language/pimpl) them, or making methods static implementation functions. The boost definitions are gathered into separate header (`mempool_set_definitions.h`) that is only included by implementation files that require definitions of the boost types. This allows us to retain the current architecture with roughly the same interfaces. 

The approach laid out by this PR also has some, albeit small, compilation speed and size benefits. Averaged over a few of compilation runs I consistently observe faster compilation by a couple of seconds and some smaller pre-processed and compiled object sizes. The main detractor of this method is obviously the number of lines touched. However it also has the benefit of inventorizing all the files that require direct access to the mempool data structures as well getting rid of boost multi index includes in non-kernel implementation files that include the mempool, but don't directly manipulate its data structures (e.g. wallet.cpp).

A much simpler alternative approach, at least on the surface, would be removing all `txmempool.h` includes from kernel library headers (see [this branch](https://github.com/TheCharlatan/bitcoin/tree/mempoolBoostSimple)). Currently this is only `validation.h`.  Due to the mutex member of `CTxMemPool` and the correspondingly defined lock decorators on the chainstate methods this becomes a bit more complicated though and I am not sure how this might be possible with the current architecture.

A discussion of how and if to remove the mempool from the kernel library has so far been intentionally punted to the next stage of the kernel library development. Pimpling the mempool itself precludes this discussion, since the library could never be shipped with the `CTxMemPool` headers. Pimpling the mempool members (like done in this PR) might also make a future splitting of block and mempool validation logic into separate compilation units easier.

sedited committed:
[refactor] Move boost txmempool definitions to their own header

Introduces mempool_set_definitions.h which collects the boost types that
were previously defined in the txmempool header.

The context of this change is an effort towards removing the boost multi
index includes from the header tree. The goal is that the experimental
`bitcoin-chainstate` binary can be compiled without requiring boost
includes.

sedited committed:
[refactor] Move boost miner definitions to miner implementation

They need not be in the header, so move them to the implementation.

sedited committed:
[refactor] Wrap txmempool's txiter in a struct

The context of this change is an effort towards removing the boost multi
index includes from the header tree. The goal is that the experimental
`bitcoin-chainstate` binary can be compiled without requiring boost
includes.

sedited committed:
[refactor] Pimpl txmempool's mapTx

The context of this change is an effort towards removing the boost multi
index includes from the header tree. The goal is that the experimental
`bitcoin-chainstate` binary can be compiled without requiring boost
includes.

sedited committed:
[refactor] Move some CTxMemPool methods to implementation.

This allows them to use the mapTx without including the full mapTx
definition in the header.

The context of this change is an effort towards removing the boost multi
index includes from the header tree. The goal is that the experimental
`bitcoin-chainstate` binary can be compiled without requiring boost
includes.

sedited committed:
[refactor] Make txmempool's GetSortedDepthAndScore static

This removes the need for defining its complex boost multi index return
type in the header.

The context of this change is an effort towards removing the boost multi
index includes from the header tree. The goal is that the experimental
`bitcoin-chainstate` binary can be compiled without requiring boost
includes.

sedited committed:
[refactor] Make txmempool's UpdateForDescendants static

The context of this change is an effort towards removing the boost multi
index includes from the header tree. The goal is that the experimental
`bitcoin-chainstate` binary can be compiled without requiring boost
includes.

sedited committed:
[refactor] Complete ripping boost includes from txmempool.h

All the header files should now no longer contain boost multi index
includes. Only the implementation files directly include the multi index
types for manipulating mempool data structures.

The context of this change is an effort towards removing the boost multi
index includes from the header tree. The goal is that the experimental
`bitcoin-chainstate` binary can be compiled without requiring boost
includes.

sedited committed:
build: Remove boost from bitcoin-chainstate CPPFLAGS

This is possible as of the previous commit.

sedited committed:
[refactor] include cmath in init.cpp

This missing include was found in the following commit.

ryanofsky commented:
This is a pretty big change, but looking over it, it seems all straightforward and mechanical, and should not be very hard to review.

I'd be interested in opinions from developers who work more with closely with mempool code about whether this improves the mempool code, or if it creates any issues.

My opinion is would be that if this improves mempool code it should be merged, and if it makes mempool code more complicated or harder to maintain, it should not be merged.

I don't think the stated goals of this PR are very compelling on their own if they are just about speeding up compilation times, or making it easier to compile the kernel library and kernel applications with different versions of boost. (I doubt there are a lot of applications that would opt for using multiple versions of boost in the same build, and if there are it would seem more straightforward for them to put an #include boundary between their application code and libbitcoin_kernel than would be is for this PR to put a boundary between mempool code and the rest of the kernel.)

But despite the large diff and long description, the changes in the PR do seem pretty straightforward, so I don't think the benefits need to be that big to justify the changes.

theuni commented:
PR title is misleading btw. Boost is not removed from the kernel, it's removed from the kernel's api.

I think conceptually [this is the commit](https://github.com/bitcoin/bitcoin/pull/28335/commits/fd49dc99398a3ec331f393f877ec9ae804373043) that really demonstrates the change here: `bitcoin-chainstate` no longer has any external dependencies.

glozow commented on src/txmempool.cpp:
```diff
@@ -53,8 +56,43 @@ bool TestLockPointValidity(CChain& active_chain, const LockPoints& lp)
+ * @param[in,out] mapTx a reference to the complete mempool map.
+ * @param[in] limits the mempool limits configuration.
+ */
+static void UpdateForDescendants(
```
e59f92518c831f0da237ac83875709b4174b55b2
Instead of making this static, perhaps you could inline this into `UpdateTransactionsFromBlock`? Dropping the thread-safety annotations is a bit of a negative imo.

glozow reviewed (commented):
> boost is now not a requirement for using libbitcoinkernel _at all_. It removes our _last_ external dependency. That's a huge achievement imo.

Sounds pretty cool to me.

I'm also quite interested in this advertised benefit:

> Pimpling the mempool members (like done in this PR) might also make a future splitting of block and mempool validation logic into separate compilation units easier.

> My opinion is would be that if this improves mempool code it should be merged, and if it makes mempool code more complicated or harder to maintain, it should not be merged.

I think the size of the diff is a symptom of lots of code external to txmempool knowing a bit too much about how txmempool is implemented, i.e. accessing `mapTx`  when it should probably use a different `CTxMemPool` method to get the information it needs. Adding `.impl` in those areas seems a bit ugly tbh - I think a cleaner approach would be to have those bits of code not use txiter/mapTx/setEntries. Apart from that, it seems generally good to move more of the implementation inward. I just had 1 comment about `UpdateForDescendants`.

[This branch](https://github.com/glozow/bitcoin/tree/2023-08-reduce-nonmempool-boost) refactors some of these `mapTx` accesses to use things like `infoAll()` and `GetIter()` and `size()`, which imo are the more appropriate functions to be calling there. The result is that a lot of things don't need to be pimpl'd anymore or fewer places are using the pimpl'd data structures. I rebased this PR on top of that branch, which resulted in 7 commits being unnecessary and the pimpling commits touching fewer files. The overall diff isn't that much smaller, but I think it achieves the intended result + independently nice changes.

I sorted the commits from most trivial to least trivial - the first ones are very straightforward things like "call pool.size() instead of pool.mapTx.size()." The last part re-implements `DisconnectedBlockTransactions` to not need a boost multi index container, which I suppose could be controversial. Would you maybe be interested in taking any of these changes?

theuni commented:
> In any case, I'm happy about the benefits of this PR, but just think the main consideration here should be whether it improves the mempool code or not, and glad to see glozow had some suggestions for simplifying it.

I think we should leave it at this. I'll admit to complete tunnel-vision on this particular boost removal. It's been ~an obsession~ a goal of mine for long enough that I'm probably too biased to judge the code changes required for it. And as you said, I don't think i need to belabor that point anyway when there's other merit to the changes.

@glozow thanks for taking a look!  Seems getting rid of some of the direct uses of `mapTx` would be a nice cleanup even outside of this PR?

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/28335#pullrequestreview-1603077161

Thank you for taking so much consideration for this change! I think your additional patches are well worth reducing the amount of code that has to directly access the boost data structure from outside of the `CTxMemPool` class.

> Would you maybe be interested in taking any of these changes?

Most of the changes (as you mentioned) should be uncontroversial, e.g. just directly using existing helpers, or the `vTxHashes` changes. I'll gladly integrate these here. 

I think the `DisconnectedBlockTransactions` change would merit its own PRs, since it does change the internal behavior and performance somewhat. 

The other two cases I am not sure about is using `infoAll()` (as done [here](https://github.com/glozow/bitcoin/commit/c287d8dc5df644b04fa9b4ffc62c8537ce76730a#diff-9c5b83de6dc84af277e352c88b9291aa44340a3c75f572a0b51661eb0a838de9R347)) to create a vector of `TxMemPoolInfo` instead of directly iterating over `mapTx` and construction of a new `GenTxid` [here](https://github.com/glozow/bitcoin/commit/8dd0cf59650ab1ed718515e2a5636d241d33637a). In both cases it would be fairly straight forward to replace them with new  `CTxMemPool` helper methods if they prove too controversial.

TheCharlatan commented:
Updated with @glozow's patches. I think these are well worth it for both cleaning up the code to not needlessly reach into complex mempool member variables as well as reducing the number of source files that need to include the multiindex headers.

The order of the commits was slightly changed, the heavier `DisconnectedBlockTransactions` refactor is now the first commit. As a further approach I think two separate PRs should be opened; one for the heavier `DisconnectedBlockTransactions` and related changes as well as another one for the easier refactors. Once these are processed, this current PR can be rebased and opened for code review.

theuni commented on src/txmempool.h:
```diff
@@ -34,6 +34,16 @@
 
 class CChain;
 
+#ifndef BITCOIN_MEMPOOL_SET_DEFINITIONS_H
```
1. I don't think this guard is necessary.
2. `DisconnectedTransactionsIteratorImpl` and `IndexedDisconnectedTransactionsImpl` don't seem to exist
3. Need to remove `mempool_set_definitions.h` to _actually_ get boost out of the headers, no?

TheCharlatan replied:
Will fixup points 2 and 3, must have slipped through the rebase.

> I don't think this guard is necessary.

The reason I added this here is that if there is an implementation that first includes the `mempool_set_definitions` and then the `txmempool` we don't redefine these data structures. Is there a better way to do this?

theuni replied:
Correct me if I'm misunderstanding, but I believe this is the pattern?
```c++
struct foo;
struct foo
{
    int bar;
};
struct foo;
int main(){}
```

If so, I (and gcc/clang :) don't see an issue with that.

theuni replied:
(Didn't mean to get in the weeds with review, I realize you're still looking for concept ACKs. The `mempool_set_definitions.h` thing is all I really care about at this stage)

TheCharlatan replied:
> If so, I (and gcc/clang :) don't see an issue with that.

I thought for some reason that this was discouraged, but you're obviously right.

> Didn't mean to get in the weeds with review,

np, any engagement here is much appreciated :)

theuni commented on src/Makefile.am:
```diff
@@ -890,7 +890,7 @@ bitcoin_util_LDADD = \
 # bitcoin-chainstate binary #
 bitcoin_chainstate_SOURCES = bitcoin-chainstate.cpp
-bitcoin_chainstate_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES) $(BOOST_CPPFLAGS)
+bitcoin_chainstate_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES)
```
FWIW this won't have any effect on c-i currently because boost is installed into the common depends prefix.

As a quick hack to demonstrate the functionality of this PR, you could add: https://github.com/theuni/bitcoin/commit/e131d8f1e33baa9a9499d2e1a4a99b171566c5a5 and drop it before merge. That puts boost in a different path, so `BOOST_CPPFLAGS` becomes a requirement as-intended. (We should also fix that for good in a less hackish way.)

As far as I can tell this PR should currently fail with the above patch because the removal is not yet done as mentioned here: https://github.com/bitcoin/bitcoin/pull/28335/commits/bdbc5093c226a8f843f938cc51e2077c4b4623a0#r1313092304

hebasto replied:
> As a quick hack to demonstrate the functionality of this PR, you could add: [theuni@e131d8f](https://github.com/theuni/bitcoin/commit/e131d8f1e33baa9a9499d2e1a4a99b171566c5a5) and drop it before merge. That puts boost in a different path, so `BOOST_CPPFLAGS` becomes a requirement as-intended.

IIRC, this issue is recurring, no?

> We should also fix that for good in a less hackish way.

Agree with it :)

theuni replied:
Heh, yes, I think this is the 3rd time we've needed this to demonstrate a boost dependency removal. Will do.

glozow commented:
> As a further approach I think two separate PRs should be opened; one for the heavier DisconnectedBlockTransactions and related changes as well as another one for the easier refactors.

> The other two cases I am not sure about is using infoAll() (as done [here](https://github.com/glozow/bitcoin/commit/c287d8dc5df644b04fa9b4ffc62c8537ce76730a#diff-9c5b83de6dc84af277e352c88b9291aa44340a3c75f572a0b51661eb0a838de9R347)) to create a vector of TxMemPoolInfo instead of directly iterating over mapTx and construction of a new GenTxid [here](https://github.com/glozow/bitcoin/commit/971133f2ed0237c9473307f292bc365c47f7de0b#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R362). In both cases it would be fairly straight forward to replace them with new CTxMemPool helper methods if they prove too controversial.

Opened #28385 with the less straightforward changes.

theuni commented:
Needs rebase after #28385.

sedited committed:
mempool: Replace sets of txiter with CTxMemPoolEntryRefs

Currently the mempool returns and consumes sets of multiindex iterators
in its public API. A likely motivation for this over working with
references to the actual values is that the multi index interface works
with these iterators and not with pointers or references to the actual
values.

However, the iterator type provides little benefit in practice as
applied currently. Its purpose, ownership, and safety semantics often
remain ambiguous, and it is hardly used for actually iterating through
the data structures. So replace it where possible with
CTxMemPoolEntryRefs.

Since CTxMemPoolEntry itself refers to its Parents and Children by
CTxMemPoolEntryRef and not txiter, this allowed for an overall
reduction of calls to iterator_to.

This also makes the long term goal of eliminating boost types from the
headers more feasible.

sedited committed:
mempool: Remove txiter/setEntries from public interface

sedited committed:
mempool: Remove dead code

TheCharlatan commented:
Rebased on #28886. which allowed for dropping one of the big commits in the previous patchset as well as slimming down the other patches.

sedited committed:
refactor: Remove duplicate CTxMemPoolEntryRef decl

TheCharlatan commented on src/txmempool.h:
```diff
@@ -396,15 +396,18 @@ class CTxMemPool
 
     typedef std::set<txiter, CompareIteratorByHash> setEntries;
 
+    typedef std::set<CTxMemPoolEntryRef, CompareIteratorByHash> setEntryRefs;
```
Added a new commit for this.

Maytow commented on src/txmempool.cpp:
```diff
@@ -12,6 +12,7 @@
 #include <consensus/tx_verify.h>
 #include <consensus/validation.h>
 #include <logging.h>
```
```suggestion
#include <logging.h>
```

Maytow reviewed (commented):
src/txmempool.cpp

sedited committed:
doc: Document CTxMemPoolEntry copy and move constructors

TheCharlatan commented:
Closing this. There is some work being done towards removing boost from the mempool wholesale. Since that would be preferable, pruning just the headers can be revisited in case that effort proves unsuccessful.
